From 59c3133a05e2df3225a4769e093fa2ede08a31e1 Mon Sep 17 00:00:00 2001
From: Stefan Berger <s.berger81@gmail.com>
Date: Sun, 15 Nov 2015 12:35:59 +0100
Subject: [PATCH 6/7] base: Add legacy methods for runtime start() and
 startVm()

These methods have been changed due to this October security patch:
https://android.googlesource.com/platform/frameworks/base.git/+/d4d3181d013f8d9b96de8e396b029f8a418894a7%5E!/

However, some frameworks (e.g. Xposed) still rely on the previous definitions,
so introduce legacy method stubs which in turn call the updated start() and
startVm() methods.

Change-Id: Ib4622ae8b475e70a12f0806daafab55077cfa0ec
---
 core/jni/AndroidRuntime.cpp              | 26 ++++++++++++++++++++++++++
 include/android_runtime/AndroidRuntime.h |  2 ++
 2 files changed, 28 insertions(+)

diff --git a/core/jni/AndroidRuntime.cpp b/core/jni/AndroidRuntime.cpp
index 35ba9d4..7e95901 100644
--- a/core/jni/AndroidRuntime.cpp
+++ b/core/jni/AndroidRuntime.cpp
@@ -436,6 +436,19 @@ void AndroidRuntime::parseExtraOpts(char* extraOptsBuf)
  *
  * Returns 0 on success.
  */
+int AndroidRuntime::startVm(JavaVM** pJavaVM, JNIEnv** pEnv)
+{
+    return startVm(pJavaVM, pEnv, false);
+}
+
+/*
+ * Start the Dalvik Virtual Machine.
+ *
+ * Various arguments, most determined by system properties, are passed in.
+ * The "mOptions" vector is updated.
+ *
+ * Returns 0 on success.
+ */
 int AndroidRuntime::startVm(JavaVM** pJavaVM, JNIEnv** pEnv, bool zygote)
 {
     int result = -1;
@@ -810,6 +823,19 @@ char* AndroidRuntime::toSlashClassName(const char* className)
  * Passes the main function two arguments, the class name and the specified
  * options string.
  */
+void AndroidRuntime::start(const char* className, const char* options)
+{
+    start(className, options, false);
+}
+
+/*
+ * Start the Android runtime.  This involves starting the virtual machine
+ * and calling the "static void main(String[] args)" method in the class
+ * named by "className".
+ *
+ * Passes the main function two arguments, the class name and the specified
+ * options string.
+ */
 void AndroidRuntime::start(const char* className, const char* options, bool zygote)
 {
     ALOGD("\n>>>>>> AndroidRuntime START %s <<<<<<\n",
diff --git a/include/android_runtime/AndroidRuntime.h b/include/android_runtime/AndroidRuntime.h
index e8e869d..fa09643 100644
--- a/include/android_runtime/AndroidRuntime.h
+++ b/include/android_runtime/AndroidRuntime.h
@@ -64,6 +64,7 @@ public:
 
     int addVmArguments(int argc, const char* const argv[]);
 
+    void start(const char *classname, const char* options);
     void start(const char *classname, const char* options, bool zygote);
 
     void exit(int code);
@@ -116,6 +117,7 @@ public:
 private:
     static int startReg(JNIEnv* env);
     void parseExtraOpts(char* extraOptsBuf);
+    int startVm(JavaVM** pJavaVM, JNIEnv** pEnv);
     int startVm(JavaVM** pJavaVM, JNIEnv** pEnv, bool zygote);
 
     Vector<JavaVMOption> mOptions;
-- 
2.5.0

